src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala

```diff
@@ -168,46 +168,64 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
     }
   }
 
-  /**
-   * Read until the given data are matched with `mat`.
-   * When withinBlock is true, it saves the data came in.
-   *
-   * @param mat bytes to match
-   * @param withinBlock start offset
-   * @return whether it finds the match successfully
-   */
-  private def readUntilMatch(mat: Array[Byte], withinBlock: Boolean): Boolean = {
-    var i: Int = 0
-    while(true) {
-      val b: Int = in.read
-      if (b == -1) {
-        currentStartTag = startTag
-        return false
-      }
-      if (withinBlock) {
-        buffer.write(b)
-      }
-      if (b == mat(i)) {
-        i += 1
-        if (i >= mat.length) {
-          currentStartTag = startTag
-          return true
+  private def readUntilStartElement(): Boolean = {
+    @annotation.tailrec
+    def loop(i: Int): Boolean = {
```

Comment:
So, personally I like the changes in general but I am just wondering if we can avoid recursion but use just while loop.